### ISSUE TITLE
Support 303 status for authorization response

### ISSUE THREAD

**Issue by wapkch** [spring-projects/spring-authorization-server#1051]
**Expected Behavior**

According to [A comprehensive formal security analysis of OAuth 2.0](https://blog.acolyer.org/2016/11/07/a-comprehensive-formal-security-analysis-of-oauth-2-0/). 303 redirect should be used to drop the body of an HTTP POST request.

**Current Behavior**

DefaultRedirectStrategy in OAuth2AuthorizationEndpointFilter sets the status to 302

**Context**

If needed, i can work on it.

**Comment by jgrandja** [comment]
@wapkch The article you referenced is dated 2016 so it's quite old and OAuth2 has gone through many revisions since.

> 303 redirect should be used to drop the body of an HTTP POST request.

Can you please provide more details on your reasoning? Do you see issues with the current implementation of `OAuth2AuthorizationEndpointFilter` returning 302? If so, please provide specific details.

**Comment by wapkch** [comment]
> @wapkch The article you referenced is dated 2016 so it's quite old and OAuth2 has gone through many revisions since.
> 
> > 303 redirect should be used to drop the body of an HTTP POST request.
> 
> Can you please provide more details on your reasoning? Do you see issues with the current implementation of `OAuth2AuthorizationEndpointFilter` returning 302? If so, please provide specific details.

In my scenario, no issue so far. And probably not a problem for most modern browsers.

But i'm afraid some user-agents treat 302 like 307 which then would expose the user's credentials to the OAuth2 client on the callback.
<img width="1358" alt="image" src="https://user-images.githubusercontent.com/32868606/214198766-df37e5ce-c8a9-4d18-970e-999d0111a3a7.png">

So, 303 is the best choice. Or at least, provide an easy way to replace the `DefaultRedirectStrategy` in OAuth2AuthorizationEndpointFilter instead of having to replacing the whole `OAuth2AuthorizationEndpointFilter`.

**Comment by jgrandja** [comment]
@wapkch I wasn't able to reproduce the issue. Your request sequence is different than the default [Messages sample](https://github.com/spring-projects/spring-authorization-server/tree/main/samples#messages-sample) as it appears you are not using OIDC for authentication.

Can you please provide a [minimal sample](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces the issue so I can replicate on my end.

**Comment by wapkch** [comment]
Hi @jgrandja, In the above scenario, I was assuming a user-agent which treats 302 like 307(Not encountered so far, but it is possible). So i changed the `RedirectStrategy`  in `SavedRequestAwareAuthenticationSuccessHandler` and `OAuth2AuthorizationEndpointFilter` from the `DefaultRedirectStrategy` to a custom `Http307RedirectStrategy` to simulate the scene. The minimal sample can be found [here](https://github.com/wapkch/spring-authorization-server-demo/tree/issues/1051). The reproduction process is as follows:

1. Start the application, default at localhost:9000
2. Authorization request
```
http://localhost:9000/oauth2/authorize?response_type=code&client_id=messaging-client&scope=openid&redirect_uri=http://127.0.0.1:8080/authorized&state=some-state
```
3. Login via form
4. OAuth2 callback request with user credentials exposed to the client like following:
<img width="1377" alt="image" src="https://user-images.githubusercontent.com/32868606/214762955-6b42ce13-fc2e-4f86-bb2d-8bc67ecc8769.png">

**Comment by jgrandja** [comment]
@wapkch Thanks for providing the sample.

The customization applied in `DefaultSecurityConfig`, specifically with the [formLogin()](https://github.com/wapkch/spring-authorization-server-demo/blob/issues/1051/src/main/java/sample/config/DefaultSecurityConfig.java#L50) authentication mechanism is what triggers the issue. That configuration is the authentication mechanism used and is out-of-scope for Spring Authorization Server. This is a Spring Security concern since authentication is configured and handled by Spring Security. Based on this alone, the issue is not specific to Spring Authorization Server.

Furthermore, I don't see this being an issue with Spring Security either. The `DefaultRedirectStrategy` calls the underlying implementation of  `HttpServletResponse.sendRedirect(location)`, which specifies that the `302` status must be returned.

> But i'm afraid some user-agents treat 302 like 307 which then would expose the user's credentials to the OAuth2 client on the callback.

> I was assuming a user-agent which treats 302 like 307(**Not encountered so far**, but it is possible)

It appears you haven't found a user-agent that actually performs this behaviour. However, if you did, this is a bug in the user-agent and should be reported to their ticketing system.

Also, consider a similar use case where `formLogin()` is configured to return a `307` status but the `location` is a simple `@Controller` (NOT an OAuth2 Client Redirection Endpoint) then it too would have access to the submitted user's credentials. Again, this scenario is not specific to OAuth.

Based on the explanation provided, I'm going to close this as I don't see any issues in Spring Authorization Server or Spring Security.

**Comment by wapkch** [comment]
Hi @jgrandja Thanks for your detailed reply. 

I agree that it's the authentication mechanism in Spring Security triggers the issue, and it's not specific to OAuth. But i think there are still some optimization we can do: 

> It appears you haven't found a user-agent that actually performs this behaviour. However, if you did, this is a bug in the user-agent and should be reported to their ticketing system.

I'm not sure if this is a bug in the user-agent. Because according to [rfc2616](https://www.rfc-editor.org/rfc/rfc2616#page-62): 

> Note: [RFC 1945](https://www.rfc-editor.org/rfc/rfc1945) and [RFC 2068](https://www.rfc-editor.org/rfc/rfc2068) specify that the client is not allowed
      to change the method on the redirected request.  However, most
      existing user agent implementations treat 302 as if it were a 303
      response, performing a GET on the Location field-value regardless
      of the original request method. The status codes 303 and 307 have
      been added for servers that wish to make unambiguously clear which
      kind of reaction is expected of the client.

 Some user-agents erroneously redirect a 302 POST into a GET request, so 303 and 307 was added to make it clear. And later in [rfc7231](https://www.rfc-editor.org/rfc/rfc7231#section-6.4.3): 

> The user agent MAY
   use the Location field value for automatic redirection.

There is no clear instructions that 302 can not be used to redirect a POST into a POST. But 303 is can do that: 

> A user agent can perform
   a retrieval request targeting that URI (a GET or HEAD request if
   using HTTP), which might also be redirected, and present the eventual
   result as an answer to the original request.  

So, changing the status from 302 to 303 for authorization response can avoid the risks. Do you agree with that?

On the other hand, do you think adding a setter method for `RedirectStrategy` in `OAuth2AuthorizationEndpointFilter` makes sense? My `Spring Security OAuth` based project returns a 303 status in authorization response, i have to make it compatiable to all integrated parties after the migration to Spring Authorization Server.

**Comment by jgrandja** [comment]
@wapkch 

> So, changing the status from 302 to 303 for authorization response can avoid the risks. Do you agree with that?

Yes. But we'll have to figure out an improvement that won't break existing applications.

> Even without modifying the default behavior, do you think adding a setter method for `RedirectStrategy` in `OAuth2AuthorizationEndpointFilter` makes sense?

I don't think I want to expose `setRedirectStrategy()` since the status code (or authorization response) could be customized by setting a custom `setAuthenticationSuccessHandler()`.

I'll re-open this and we can consider what the best option would be.
